File: togo/internal/cli/cache.go
package cli

import (
	"fmt"
	"os"

	"github.com/nikeee/cappu/internal/cache"
	"github.com/nikeee/cappu/internal/install"
)

// RunCache handles `cappu cache clean` (remove the global download cache) and
// `cappu cache verify` (check cached artifacts against the hashes recorded
// beside them). Other subcommands are rejected. Port of src/cli/cache.ts.
func RunCache(args []string) int {
	if len(args) == 1 && args[0] == "clean" {
		removed := cache.Clean()
		if len(removed) == 0 {
			fmt.Fprintln(os.Stderr, "cache already empty")
		} else {
			for _, dir := range removed {
				fmt.Fprintf(os.Stdout, "removed %s\n", dir)
			}
		}
		return 0
	}
	if len(args) == 1 && args[0] == "verify" {
		result := install.VerifyCache()
		for _, f := range result.Modified {
			fmt.Fprintf(os.Stderr, "error: %s: cached bytes do not match the recorded hash\n", f)
		}
		for _, f := range result.Missing {
			fmt.Fprintf(os.Stderr, "error: %s: a hash is recorded but the file is gone\n", f)
		}
		fmt.Fprintf(os.Stderr, "%d ok, %d modified, %d missing\n",
			len(result.OK), len(result.Modified), len(result.Missing))
		if len(result.Modified)+len(result.Missing) > 0 {
			return 1
		}
		return 0
	}
	fmt.Fprintln(os.Stderr, "usage: cappu cache <clean|verify>")
	return 2
}

File: src/cli/cache.ts
// `cappu cache clean`: remove the global download cache (packages, JDKs,
// resolved metadata). `cappu cache verify`: check the cached artifacts against
// the hashes recorded beside them. Other `cache` subcommands are rejected.

import { cleanCache } from "../cacheDir.ts";
import { verifyCache } from "../install.ts";

export function runCacheCommand(args: readonly string[]): never {
  if (args.length === 1 && args[0] === "clean") {
    let removed: string[];
    try {
      removed = cleanCache();
    } catch (e) {
      // An undeletable cache dir is a real error, not a stack trace (and not
      // the Go build's former silent success).
      process.stderr.write(`cappu: ${(e as Error).message}\n`);
      process.exit(1);
    }
    if (removed.length === 0) {
      process.stderr.write("cache already empty\n");
    } else {
      for (const dir of removed) process.stdout.write(`removed ${dir}\n`);
    }
    process.exit(0);
  }
  if (args.length === 1 && args[0] === "verify") {
    const result = verifyCache();
    for (const file of result.modified)
      process.stderr.write(`error: ${file}: cached bytes do not match the recorded hash\n`);
    for (const file of result.missing)
      process.stderr.write(`error: ${file}: a hash is recorded but the file is gone\n`);
    process.stderr.write(
      `${result.ok.length} ok, ${result.modified.length} modified, ${result.missing.length} missing\n`,
    );
    process.exit(result.modified.length + result.missing.length > 0 ? 1 : 0);
  }
  process.stderr.write("usage: cappu cache <clean|verify>\n");
  process.exit(2);
}
